compiler/rustc_lint/src/default_could_be_derived.rs

```diff
@@ -0,0 +1,422 @@
+                                    hir::StructTailExpr::None(span) => {
+                                        sugg.push((span, ", ..".to_string()));
+                                    }
+                                }
+                                diag.multipart_suggestion_verbose(
+                                    "use the default values in the `impl` to avoid them diverging \
+                                     over time",
+                                    sugg,
+                                    Applicability::MachineApplicable,
+                                );
+                            },
+                        );
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+fn check_path<'tcx>(
+    cx: &LateContext<'tcx>,
+    path: &hir::QPath<'_>,
+    hir_id: hir::HirId,
+    ty: Ty<'tcx>,
+) -> bool {
+    let Some(default_def_id) = cx.tcx.get_diagnostic_item(sym::Default) else {
+        return false;
+    };
+    let res = cx.qpath_res(&path, hir_id);
+    let Some(def_id) = res.opt_def_id() else { return false };
+    if cx.tcx.is_diagnostic_item(sym::default_fn, def_id) {
+        // We have `field: Default::default(),`. This is what the derive would do already.
+        return true;
+    }
+    // For every `Default` impl for this type (there should be a single one), we see if it
+    // has a "canonical" `DefId` for a fn call with no arguments, or a path. If it does, we
+    // check that `DefId` with the `DefId` of this field's value if it is also a call/path.
+    // If there's a match, it means that the contents of that type's `Default` impl are the
+    // same to what the user wrote on *their* `Default` impl for this field.
+    let mut equivalents = vec![];
+    cx.tcx.for_each_relevant_impl(default_def_id, ty, |impl_def_id| {
+        let equivalent = match impl_def_id.as_local() {
+            None => cx.tcx.get_default_impl_equivalent(impl_def_id),
+            Some(local) => {
+                let def_kind = cx.tcx.def_kind(impl_def_id);
+                cx.tcx.get_default_equivalent(def_kind, local)
+            }
+        };
+        if let Some(did) = equivalent {
+            equivalents.push(did);
+        }
+    });
+    for did in equivalents {
+        if did == def_id {
+            return true;
+        }
+    }
+    false
+}
+
+fn check_expr(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+    match expr.kind {
+        hir::ExprKind::Lit(spanned_lit) => match spanned_lit.node {
+            LitKind::Int(val, _) if val == 0 => true, // field: 0,
+            LitKind::Bool(false) => true,             // field: false,
+            _ => false,
+        },
+        hir::ExprKind::Call(hir::Expr { kind: hir::ExprKind::Path(path), hir_id, .. }, []) => {
+            // `field: foo(),` or `field: Ty::assoc(),`
+            let Some(ty) = cx
+                .tcx
+                .has_typeck_results(expr.hir_id.owner.def_id)
```

Comment:
Replace this and th other calls to `has_typeck_results` with `cx.typeck_results()` which is guaranteed to work if you're already in a visitor with expressions.